README.md
seq2gif
=======
[![Build Status](https://travis-ci.org/saitoha/seq2gif.svg?branch=master)](https://travis-ci.org/saitoha/seq2gif)
[![Coverage Status](https://img.shields.io/coveralls/saitoha/seq2gif.svg)](https://coveralls.io/r/saitoha/seq2gif)


Description
------------

This software converts a sequences record file generated by [ttyrec](https://github.com/mjording/ttyrec)
into a gif animation directly(without taking screenshots),
by using portable built-in terminal emulation engine originated
from [yaft](https://github.com/uobikiemukot/yaft).
[yaft](https://github.com/uobikiemukot/yaft) provides rare terminal emulation
features such as SIXEL/DRCS.


Demo
----

Play [sl](https://github.com/mtoyoda/sl).

![img2sixel](https://raw.githubusercontent.com/saitoha/seq2gif/master/images/sl.gif)


Show SIXEL animation with [img2sixel](https://github.com/saitoha/libsixel#img2sixel).

![img2sixel](https://raw.githubusercontent.com/saitoha/seq2gif/master/images/sixel.gif)


The following demo plays [sl](https://github.com/mtoyoda/sl) and records it with [ttyrec](https://github.com/mjording/ttyrec), and
converts that record into a GIF animation using seq2gif,
and converts that GIF into a SIXEL animation and plays it with [img2sixel](https://github.com/saitoha/libsixel#img2sixel).
Furthermore, generates the log record of above experience using [ttyrec](https://github.com/mjording/ttyrec), and
converts that log into the following GIF animation.

![ttyrec](https://raw.githubusercontent.com/saitoha/seq2gif/master/images/ttyrec.gif)


Download
--------
Source package and Windows binary is [available](https://github.com/saitoha/seq2gif/releases/).


Build and Install
-----------------

This software works in Windows, Linux, BSDs and OSX.
If you want to build windows native build, [MinGW](http://www.mingw.org/) is required.

```
  $ ./configure
  $ make
  # make install
```

Usage
-----

```
Usage: seq2gif [Options] < ttyrecord > record.gif
       seq2gif [Options] -i ttyrecord -o record.gif

Options:
-w WIDTH, --width=WIDTH               specify terminal width in cell size.
                                      (default: 80)
-h HEIGHT, --height=HEIGHT            specify terminal height in cell size.
                                      (default: 24)
-l DELAY, --last-frame-delay=DELAY    specify delay in msec which is added
                                      to the last frame. (default: 300)
-f COLORNO --foreground-color=COLORNO specify foreground color palette.
                                      number.
-b COLORNO --background-color=COLORNO specify background color palette
                                      number.
-c COLORNO --cursor-color=COLORNO     specify cursor color palette
                                      number.
-t TABSTOP --tabstop=TABSTOP          specify hardware tabstop(default: 8)
-j --cjkwidth                         treat East Asian Ambiguous width
                                      characters (UAX#11) as wide.
-r COUNT --repeat=COUNT               specify animation repeat count. loop
                                      infinitely if 0 is given. (default: 0)
-i FILE --input=FILE                  specify input file name. use STDIN
                                      if '-' is given. (default: '-')
-o FILE --output=FILE                 specify output file name. use STDOUT
                                      if '-' is given. (default: '-')
-V, --version                         show version and license information.
-H, --help                            show this help.
```


Acknowledgements
----------------

### recterm

This program is derived from [recterm](https://github.com/uobikiemukot/recterm)
written by haru <uobikiemukot at gmail dot com> ([@uobikiemukot](https://github.com/uobikiemukot/))

### yaft

Some part of glyph data generator(glyph/*) is imported from [yaft](http://uobikiemukot.github.io/yaft/)
written by haru <uobikiemukot at gmail dot com> ([@uobikiemukot](https://github.com/uobikiemukot/))

### gifsave89

GIF encoder (gifsave89.c) is imported from [gifsave89](http://www.forkosh.com/gifsave90.html)
distributed under GPLv3+ by [John Forkosh Associates, Inc.](http://www.forkosh.com)(john@forkosh.com)

### milkjf
This package includes a milkjf font and its derivative(gryph/milkjf.h).
Original milkjf font was designed by japanfarm (Nihon-Nouen) and used on Sharp X68000 series.
Chisato Yamauchi converted milkjf font to bdf for X Window System.
haru([@uobikiemukot](https://github.com/uobikiemukot/)) converted it to a C-styled header file.

### M+ BITMAP FONTS
This package includes [M+ Bitmap Font](http://mplus-fonts.sourceforge.jp/mplus-bitmap-fonts/index.html)
and its derivative(gryph/mplus.h).

License:

```
M+ BITMAP FONTS            Copyright 2002-2005  COZ <coz@users.sourceforge.jp>

LICENSE

These fonts are free softwares.
Unlimited permission is granted to use, copy, and distribute it, with
or without modification, either commercially and noncommercially.
THESE FONTS ARE PROVIDED "AS IS" WITHOUT WARRANTY.
```

### GNU Unifont
This package includes [GNU Unifont Glyphs](http://unifoundry.com/unifont.html)
and its derivative(gryph/unifont.h).

License: GPLv2+
http://unifoundry.com/LICENSE.txt


```
M+ BITMAP FONTS            Copyright 2002-2005  COZ <coz@users.sourceforge.jp>

LICENSE

These fonts are free softwares.
Unlimited permission is granted to use, copy, and distribute it, with
or without modification, either commercially and noncommercially.
THESE FONTS ARE PROVIDED "AS IS" WITHOUT WARRANTY.
```

### M+ BITMAP FONTS
This package includes a C header file derived from M+ font(gryph/mplus.h).


### Markus Kuhn's wcwidth
This package includes Markus Kuhn's wcwidth(mk_wcwidth)

```
Markus Kuhn -- 2007-05-26 (Unicode 5.0)

Permission to use, copy, modify, and distribute this software
for any purpose and without fee is hereby granted. The author
disclaims all warranties with regard to this software.

Latest version: http://www.cl.cam.ac.uk/~mgk25/ucs/wcwidth.c
```

Contributors
-------------

- [@arakiken](https://bitbucket.org/arakiken/profile/repositories)
- [IWAMOTO Kouichi(@ttdoda)](https://github.com/ttdoda)


Similar Software
----------------

### [sugyan/ttyrec2gif](https://github.com/sugyan/ttyrec2gif)

This also generates GIF animation without taking screenshots, written in [Go](http://golang.org/).
Using [j4k.co/terminal](http://godoc.org/j4k.co/terminal) as terminal emulation engine().
It is derived from [sugyan/ttygif](https://github.com/sugyan/tty2gif).


### [uobikiemukot/recterm/tools/rec2gif.c](https://github.com/uobikiemukot/recterm/blob/master/tools/rec2gif.c)

Additional tools which uses [yaft](https://github.com/uobikiemukot/yaft) terminal emulation engine,
distributed with [uobikiemukot/recterm](https://github.com/uobikiemukot/recterm/).


References
----------

### [ttyrec](http://0xcc.net/ttyrec/index.html.en)

A tty recorder.

### [ttyrec4windows](https://github.com/mattn/ttyrec4windows)

A Windows port for [ttyrec](http://0xcc.net/ttyrec/index.html.en) written in [Go](http://golang.org/).


License
-------

![GPLv3](https://raw.githubusercontent.com/saitoha/seq2gif/master/images/gplv3.png)



ChangeLog
2014-08-01  Hayaki Saito <user@zuse.jp>

  * README.md, images/sixel.gif, images/sl.gif, tests/data/sixel.tty: Add a
  test record for sixel

2014-07-31  Hayaki Saito <user@zuse.jp>

  * .coveralls.yml, .travis.yml, Makefile.am, Makefile.in, main.c: Add an error
  checking code

  * Makefile.am, Makefile.in, configure, configure.ac: Add --enable-gcov
  configure option

  * dcs.c, pseudo.c: Strip inline keywords

  * Makefile.am, Makefile.in: Check memory violation errors with valgrind

  * Makefile.am, Makefile.in: Add "valgrind" makefile target

  * configure, configure.ac, main.c, package.json: Fix memory access violation

  * ChangeLog, configure, configure.ac, package.json: Update minor version

  * tests/data/ls.tty, tests/data/sl.tty: Add test data

  * config.h.in, configure, configure.ac, yaft.h: Add --with-fontdata configure
  option

  * configure, configure.ac, package.json: Bump version

  * package.json, package.json.in: Add missing files

  * Makefile.in, main.c: Add --repeat option

  * Makefile.in, configure, configure.ac: Add clib integration

  * README.md, main.c: Minor fix

  * configure, configure.ac: Update minor version

  * README.md, main.c, terminal.c, terminal.h, yaft.h: Add --cjkwidth option

2014-07-30  Hayaki Saito <user@zuse.jp>

  * configure, configure.ac, main.c: Update micro version

  * config.h.in, configure, configure.ac, main.c: Minor fix

  * config.h.in, configure, configure.ac, main.c: Fix for mingw environment

  * configure, configure.ac: Update version

  * Makefile.am, Makefile.in, malloc_stub.c, malloc_stub.h, terminal.c,
  wcwidth.h, yaft.h: Minor fix

2014-07-29  Hayaki Saito <user@zuse.jp>

  * config.h.in, configure, configure.ac, util.c: Checks calloc availability

  * README.md, dcs.c: Strip a needless strlen

  * README.md, configure, configure.ac: Update README

  * README.md, main.c: Update README and output of --help option

2014-07-28  Hayaki Saito <user@zuse.jp>

  * config.h.in, configure, configure.ac, dcs.c, gifsave89.c, main.c, osc.c,
  parse.c, util.c: Update configure script

  * configure, configure.ac: Bump minor version

  * conf.h, dcs.c, function.c, main.c, osc.c, parse.c, pseudo.c, terminal.c,
  util.c, yaft.h: Drop conf.h

  * conf.h, main.c, terminal.c, terminal.h, yaft.h: Add tabstop option

  * conf.h, main.c, pseudo.c, terminal.c, terminal.h, yaft.h: Add
  --cursor-color option

  * configure, configure.ac: Bump micro version

  * main.c, parse.c, parse.h: Support sixel animation

  * configure, configure.ac: Bump version

  * README.md, conf.h, function.c, main.c, pseudo.c, terminal.c, terminal.h,
  yaft.h: Add --foreground-color/--background-color option

  * configure, configure.ac: Bump version

  * README.md, main.c: Add --last-frame-delay option

2014-07-27  Hayaki Saito <user@zuse.jp>

  * config.h.in, configure, configure.ac, dcs.c, function.c, gifsave89.h,
  main.c, osc.c, parse.c, pseudo.c, terminal.c, util.c, wcwidth.h, yaft.h:
  Include config.h in each source files

  * COPYING, INSTALL: Add proper COPYING and INSTALL file

  * Makefile.in, aclocal.m4, configure, configure.ac: Update minor version

2014-07-27  saitoha <user@zuse.jp>

  * Makefile.am, Makefile.in, aclocal.m4, config.h.in, configure, configure.ac,
  main.c, malloc_stub.c, malloc_stub.h, osc.c, terminal.c, wcwidth.c,
  wcwidth.h, yaft.h: Add malloc stub and wcwidth function

2014-07-27  Hayaki Saito <user@zuse.jp>

  * configure, configure.ac: Add a hack against that git does not preverve file
  timestamp

  * configure, configure.ac: Bump micro version

  * README.md, main.c: Update default width and height value

  * configure, configure.ac: Bump micro version

  * config.h.in, configure, configure.ac, main.c: Check getopt_long
  availability

  * configure, configure.ac, gifsave89.h, main.c: Add --width, --height,
  --version option

  * main.c, parse.c, pseudo.c, terminal.c, terminal.h, util.c, util.h, yaft.h:
  Delete unused functions and definitions

  * color.h, conf.h, dcs.c, dcs.h, function.c, function.h, main.c, osc.c,
  osc.h, parse.c, parse.h, pseudo.c, pseudo.h, terminal.c, terminal.h, util.c,
  util.h: Add mode lines

  * color.h, conf.h, dcs.c, dcs.h, function.c, function.h, gifsave89.c,
  gifsave89.h, glyph.h, glyph/milkjf.h, glyph/mplus.h, images/gplv3.png,
  main.c, osc.c, osc.h, parse.c, parse.h, pseudo.c, pseudo.h, sample/256.gif,
  sample/man.gif, sample/netlab.gif, sample/peco.gif, sample/sl.gif,
  terminal.c, terminal.h, util.c, util.h, yaft.h: Add license blocks

2014-07-26  Hayaki Saito <user@zuse.jp>

  * COPYING, INSTALL, LICENSE, Makefile.am, Makefile.in, aclocal.m4, compile,
  config.h.in, configure, configure.ac, depcomp, install-sh, main.c, makefile,
  missing: Generate configure script using autotools

2014-07-25  Hayaki Saito <user@zuse.jp>

  * dcs.c, function.c, gifsave89.c, gifsave89.h, glyph.h, main.c, osc.c,
  parse.c, pseudo.c, terminal.c, util.c, yaft.h: Add missing files

  * color.h, conf.h, gifsave89.h, util.h: Style fix

  * README.md, color.h, conf.h, dcs.h, draw.h, function.h, gifsave89.h,
  makefile, osc.h, parse.h, pseudo.h, recterm.c, recterm.h, terminal.h, util.h:
  Strip pseudo terminal and read stdin as output sequences

  * README.md, conf.h: Forked by Hayaki Saito

2014-07-25  uobikiemukot <uobikiemukot@gmail.com>

  * README.md, color.h, conf.h, dcs.h, draw.h, function.h, gifsave89.h,
  glyph/milkjf.h, glyph/mplus.h, makefile, osc.h, parse.h, pseudo.h, recterm.c,
  recterm.h, sample/256.gif, sample/man.gif, sample/netlab.gif,
  sample/peco.gif, sample/sl.gif, terminal.h, util.h: first version


main.c
/*
 * Copyright (C) 2014 haru <uobikiemukot at gmail dot com>
 * Copyright (C) 2014 Hayaki Saito <user@zuse.jp>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"
#include "yaft.h"
#include "util.h"
#include "pseudo.h"
#include "terminal.h"
#include "function.h"
#include "osc.h"
#include "dcs.h"
#include "parse.h"
#include "gifsave89.h"

#include <stdio.h>

#if HAVE_STDLIB_H
# include <stdlib.h>
#endif
#if HAVE_STRING_H
# include <string.h>
#endif
#if HAVE_UNISTD_H
# include <unistd.h>
#endif
#if HAVE_GETOPT_H
# include <getopt.h>
#endif
#if HAVE_ERRNO_H
# include <errno.h>
#endif
#if HAVE_FCNTL_H
# include <fcntl.h>
#endif


#if !defined(HAVE_MEMCPY)
# define memcpy(d, s, n) (bcopy ((s), (d), (n)))
#endif

#if !defined(O_BINARY) && defined(_O_BINARY)
# define O_BINARY _O_BINARY
#endif  /* !defined(O_BINARY) && !defined(_O_BINARY) */

struct settings_t {
    int width;
    int height;
    int show_version;
    int show_help;
    int last_frame_delay;
    int foreground_color;
    int background_color;
    int cursor_color;
    int tabwidth;
    int cjkwidth;
    int repeat;
    char *input;
    char *output;
};

enum cmap_bitfield {
    RED_SHIFT   = 5,
    GREEN_SHIFT = 2,
    BLUE_SHIFT  = 0,
    RED_MASK    = 3,
    GREEN_MASK  = 3,
    BLUE_MASK   = 2
};

static void pb_init(struct pseudobuffer *pb, int width, int height)
{
    pb->width  = width;
    pb->height = height;
    pb->bytes_per_pixel = BYTES_PER_PIXEL;
    pb->line_length = pb->width * pb->bytes_per_pixel;
    pb->buf = ecalloc(pb->width * pb->height, pb->bytes_per_pixel);
}

static void pb_die(struct pseudobuffer *pb)
{
    free(pb->buf);
}

static void set_colormap(int colormap[COLORS * BYTES_PER_PIXEL + 1])
{
    int i, ci, r, g, b;
    uint8_t index;

    /* colormap: terminal 256color
    for (i = 0; i < COLORS; i++) {
        ci = i * BYTES_PER_PIXEL;

        r = (color_list[i] >> 16) & bit_mask[8];
        g = (color_list[i] >> 8)  & bit_mask[8];
        b = (color_list[i] >> 0)  & bit_mask[8];

        colormap[ci + 0] = r;
        colormap[ci + 1] = g;
        colormap[ci + 2] = b;
    }
    */

    /* colormap: red/green: 3bit blue: 2bit
    */
    for (i = 0; i < COLORS; i++) {
        index = (uint8_t) i;
        ci = i * BYTES_PER_PIXEL;

        r = (index >> RED_SHIFT)   & bit_mask[RED_MASK];
        g = (index >> GREEN_SHIFT) & bit_mask[GREEN_MASK];
        b = (index >> BLUE_SHIFT)  & bit_mask[BLUE_MASK];

        colormap[ci + 0] = r * bit_mask[BITS_PER_BYTE] / bit_mask[RED_MASK];
        colormap[ci + 1] = g * bit_mask[BITS_PER_BYTE] / bit_mask[GREEN_MASK];
        colormap[ci + 2] = b * bit_mask[BITS_PER_BYTE] / bit_mask[BLUE_MASK];
    }
    colormap[COLORS * BYTES_PER_PIXEL] = -1;
}

static uint32_t pixel2index(uint32_t pixel)
{
    /* pixel is always 24bpp */
    uint32_t r, g, b;

    /* split r, g, b bits */
    r = (pixel >> 16) & bit_mask[8];
    g = (pixel >> 8)  & bit_mask[8];
    b = (pixel >> 0)  & bit_mask[8];

    /* colormap: terminal 256color
    if (r == g && r == b) { // 24 gray scale
        r = 24 * r / COLORS;
        return 232 + r;
    }                       // 6x6x6 color cube

    r = 6 * r / COLORS;
    g = 6 * g / COLORS;
    b = 6 * b / COLORS;

    return 16 + (r * 36) + (g * 6) + b;
    */

    /* colormap: red/green: 3bit blue: 2bit
    */
    // get MSB ..._MASK bits
    r = (r >> (8 - RED_MASK))   & bit_mask[RED_MASK];
    g = (g >> (8 - GREEN_MASK)) & bit_mask[GREEN_MASK];
    b = (b >> (8 - BLUE_MASK))  & bit_mask[BLUE_MASK];

    return (r << RED_SHIFT) | (g << GREEN_SHIFT) | (b << BLUE_SHIFT);
}

static void apply_colormap(struct pseudobuffer *pb, unsigned char *img)
{
    int w, h;
    uint32_t pixel = 0;

    for (h = 0; h < pb->height; h++) {
        for (w = 0; w < pb->width; w++) {
            memcpy(&pixel, pb->buf + h * pb->line_length
                + w * pb->bytes_per_pixel, pb->bytes_per_pixel);
            *(img + h * pb->width + w) = pixel2index(pixel) & bit_mask[BITS_PER_BYTE];
        }
    }
}

static size_t write_gif(unsigned char *gifimage, int size, FILE *f)
{
    size_t wsize = 0;

    wsize = fwrite(gifimage, sizeof(unsigned char), size, f);
    return wsize;
}

static void show_version()
{
    printf(PACKAGE_NAME " " PACKAGE_VERSION "\n"
           "Copyright (C) 2014 haru <uobikiemukot at gmail dot com>\n"
           "Copyright (C) 2012-2014 Hayaki Saito <user@zuse.jp>.\n"
           "\n"
           "This program is free software; you can redistribute it and/or modify\n"
           "it under the terms of the GNU General Public License as published by\n"
           "the Free Software Foundation; either version 3 of the License, or\n"
           "(at your option) any later version.\n"
           "\n"
           "This program is distributed in the hope that it will be useful,\n"
           "but WITHOUT ANY WARRANTY; without even the implied warranty of\n"
           "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n"
           "GNU General Public License for more details.\n"
           "\n"
           "You should have received a copy of the GNU General Public License\n"
           "along with this program. If not, see http://www.gnu.org/licenses/.\n"
           "%s\n", copyright
          );
}

static void show_help()
{
    fprintf(stderr,
            "Usage: seq2gif [Options] < ttyrecord > record.gif\n"
            "       seq2gif [Options] -i ttyrecord -o record.gif\n"
            "\n"
            "Options:\n"
            "-w WIDTH, --width=WIDTH               specify terminal width in cell size.\n"
            "                                      (default: 80)\n"
            "-h HEIGHT, --height=HEIGHT            specify terminal height in cell size.\n"
            "                                      (default: 24)\n"
            "-l DELAY, --last-frame-delay=DELAY    specify delay in msec which is added\n"
            "                                      to the last frame. (default: 300)\n"
            "-f COLORNO --foreground-color=COLORNO specify foreground color palette.\n"
            "                                      number.\n"
            "-b COLORNO --background-color=COLORNO specify background color palette\n"
            "                                      number.\n"
            "-c COLORNO --cursor-color=COLORNO     specify cursor color palette\n"
            "                                      number.\n"
            "-t TABSTOP --tabstop=TABSTOP          specify hardware tabstop(default: 8)\n"
            "-j --cjkwidth                         treat East Asian Ambiguous width\n"
            "                                      characters (UAX#11) as wide.\n"
            "-r COUNT --repeat=COUNT               specify animation repeat count. loop\n"
            "                                      infinitely if 0 is given. (default: 0)\n"
            "-i FILE --input=FILE                  specify input file name. use STDIN\n"
            "                                      if '-' is given. (default: '-')\n"
            "-o FILE --output=FILE                 specify output file name. use STDOUT\n"
            "                                      if '-' is given. (default: '-')\n"
            "-V, --version                         show version and license information.\n"
            "-H, --help                            show this help.\n"
           );
}

static int parse_args(int argc, char *argv[], struct settings_t *psettings)
{
    int long_opt;
    int n;
    char const *optstring = "w:h:HVl:f:b:c:t:jr:i:o:";
#if HAVE_GETOPT_LONG
    int option_index;
#endif  /* HAVE_GETOPT_LONG */

#if HAVE_GETOPT_LONG
    struct option long_options[] = {
        {"width",             required_argument,  &long_opt, 'w'},
        {"height",            required_argument,  &long_opt, 'h'},
        {"last-frame-delay",  required_argument,  &long_opt, 'l'},
        {"foreground-color",  required_argument,  &long_opt, 'f'},
        {"background-color",  required_argument,  &long_opt, 'b'},
        {"cursor-color",      required_argument,  &long_opt, 'c'},
        {"tabstop",           required_argument,  &long_opt, 't'},
        {"cjkwidth",          no_argument,        &long_opt, 'j'},
        {"repeat",            required_argument,  &long_opt, 'r'},
        {"input",             required_argument,  &long_opt, 'i'},
        {"output",            required_argument,  &long_opt, 'o'},
        {"help",              no_argument,        &long_opt, 'H'},
        {"version",           no_argument,        &long_opt, 'V'},
        {0, 0, 0, 0}
    };
#endif  /* HAVE_GETOPT_LONG */

    for (;;) {

#if HAVE_GETOPT_LONG
        n = getopt_long(argc, argv, optstring,
                        long_options, &option_index);
#else
        n = getopt(argc, argv, optstring);
#endif  /* HAVE_GETOPT_LONG */
        if (n == -1) {
            break;
        }
        if (n == 0) {
            n = long_opt;
        }
        switch(n) {
        case 'w':
            psettings->width = atoi(optarg);
            if (psettings->width < 1) {
                goto argerr;
            }
            break;
        case 'h':
            psettings->height = atoi(optarg);
            if (psettings->height < 1) {
                goto argerr;
            }
            break;
        case 'l':
            psettings->last_frame_delay = atoi(optarg);
            if (psettings->last_frame_delay < 0) {
                goto argerr;
            }
            break;
        case 'f':
            psettings->foreground_color = atoi(optarg);
            if (psettings->foreground_color < 0) {
                goto argerr;
            }
            if (psettings->foreground_color > 255) {
                goto argerr;
            }
            break;
        case 'b':
            psettings->background_color = atoi(optarg);
            if (psettings->background_color < 0) {
                goto argerr;
            }
            if (psettings->background_color > 255) {
                goto argerr;
            }
            break;
        case 'c':
            psettings->cursor_color = atoi(optarg);
            if (psettings->cursor_color < 0) {
                goto argerr;
            }
            if (psettings->cursor_color > 255) {
                goto argerr;
            }
            break;
        case 't':
            psettings->tabwidth = atoi(optarg);
            if (psettings->tabwidth < 0) {
                goto argerr;
            }
            if (psettings->tabwidth > 255) {
                goto argerr;
            }
            break;
        case 'j':
            psettings->cjkwidth = 1;
            break;
        case 'r':
            psettings->repeat = atoi(optarg);
            if (psettings->repeat < 0) {
                goto argerr;
            }
            if (psettings->repeat > 0xffff) {
                goto argerr;
            }
            break;
        case 'i':
            psettings->input = strdup(optarg);
            break;
        case 'o':
            psettings->output = strdup(optarg);
            break;
        case 'H':
            psettings->show_help = 1;
            break;
        case 'V':
            psettings->show_version = 1;
            break;
        default:
            goto argerr;
        }
    }
    return 0;

argerr:
    show_help();
    return 1;
}

static FILE * open_input_file(char const *filename)
{
    FILE *f;

    if (filename == NULL || strcmp(filename, "-") == 0) {
        /* for windows */
#if defined(O_BINARY)
# if HAVE__SETMODE
        _setmode(fileno(stdin), O_BINARY);
# elif HAVE_SETMODE
        setmode(fileno(stdin), O_BINARY);
# endif  /* HAVE_SETMODE */
#endif  /* defined(O_BINARY) */
        return stdin;
    }
    f = fopen(filename, "rb");
    if (!f) {
#if _ERRNO_H
        fprintf(stderr, "fopen('%s') failed.\n" "reason: %s.\n",
                filename, strerror(errno));
#endif  /* HAVE_ERRNO_H */
        return NULL;
    }
    return f;
}

static FILE * open_output_file(char const *filename)
{
    FILE *f;

    if (filename == NULL || strcmp(filename, "-") == 0) {
        /* for windows */
#if defined(O_BINARY)
# if HAVE__SETMODE
        _setmode(fileno(stdout), O_BINARY);
# elif HAVE_SETMODE
        setmode(fileno(stdout), O_BINARY);
# endif  /* HAVE_SETMODE */
#endif  /* defined(O_BINARY) */
        return stdout;
    }
    f = fopen(filename, "wb");
    if (!f) {
#if _ERRNO_H
        fprintf(stderr, "fopen('%s') failed.\n" "reason: %s.\n",
                filename, strerror(errno));
#endif  /* HAVE_ERRNO_H */
        return NULL;
    }
    return f;
}

static int32_t readtime(FILE *f, uint8_t *obuf)
{
    int nread;
    int32_t tv_sec;
    int32_t tv_usec;

    nread = fread(obuf, 1, sizeof(tv_sec), f);
    if (nread != sizeof(tv_sec)) {
        return -1;
    }
    tv_sec = obuf[0] | obuf[1] << 8
           | obuf[2] << 16 | obuf[3] << 24;
    nread = fread(obuf, 1, sizeof(tv_usec), f);
    if (nread != sizeof(tv_usec)) {
        return -1;
    }
    tv_usec = obuf[0] | obuf[1] << 8
            | obuf[2] << 16 | obuf[3] << 24;

    return tv_sec * 1000000 + tv_usec;
}

static int32_t readlen(FILE *f, uint8_t *obuf)
{
    int nread;
    uint32_t len;

    nread = fread(obuf, 1, sizeof(len), f);
    if (nread != sizeof(len)) {
        return -1;
    }
    len = obuf[0]
        | obuf[1] << 8
        | obuf[2] << 16
        | obuf[3] << 24;

    return len;
}

int main(int argc, char *argv[])
{
    uint8_t *obuf;
    ssize_t nread;
    struct terminal term;
    struct pseudobuffer pb;
    uint32_t prev = 0;
    uint32_t now = 0;
    ssize_t len = 0;
    size_t maxlen = 0;
    int delay = 0;
    int dirty = 0;
    FILE *in_file = NULL;
    FILE *out_file = NULL;
    int nret = EXIT_SUCCESS;

    void *gsdata;
    unsigned char *gifimage = NULL;
    int gifsize, colormap[COLORS * BYTES_PER_PIXEL + 1];
    unsigned char *img;

    struct settings_t settings = {
        80,     /* width */
        24,     /* height */
        0,      /* show_version */
        0,      /* show_help */
        300,    /* last_frame_delay */
        7,      /* foreground_color */
        0,      /* background_color */
        2,      /* cursor_color */
        8,      /* tabwidth */
        0,      /* cjkwidth */
        0,      /* repeat */
        NULL,   /* input */
        NULL,   /* output */
    };

    if (parse_args(argc, argv, &settings) != 0) {
        exit(1);
    }

    if (settings.show_help) {
        show_help();
        exit(0);
    }

    if (settings.show_version) {
        show_version();
        exit(0);
    }

    /* init */
    pb_init(&pb, settings.width * CELL_WIDTH, settings.height * CELL_HEIGHT);
    term_init(&term, pb.width, pb.height,
              settings.foreground_color,
              settings.background_color,
              settings.cursor_color,
              settings.tabwidth,
              settings.cjkwidth);

    /* init gif */
    img = (unsigned char *) ecalloc(pb.width * pb.height, 1);
    set_colormap(colormap);
    if (!(gsdata = newgif((void **) &gifimage, pb.width, pb.height, colormap, 0))) {
        free(img);
        term_die(&term);
        pb_die(&pb);
        return EXIT_FAILURE;
    }

    animategif(gsdata, /* repetitions */ settings.repeat, 0,
        /* transparent background */  -1, /* disposal */ 2);

    in_file = open_input_file(settings.input);
    out_file = open_output_file(settings.output);

    maxlen = 2048;
    obuf = malloc(maxlen);
    prev = now = readtime(in_file, obuf);

    /* main loop */
    for(;;) {
        len = readlen(in_file, obuf);
        if (len <= 0) {
            nret = EXIT_FAILURE;
            break;
        }
        if (len > maxlen) {
            obuf = realloc(obuf, len);
            maxlen = len;
        }
        nread = fread(obuf, 1, len, in_file);
        if (nread != len) {
            nret = EXIT_FAILURE;
            break;
        }
        parse(&term, obuf, nread, &dirty);
        now = readtime(in_file, obuf);
        if (now == -1) {
            break;
        }
        if (term.esc.state != STATE_DCS || dirty) {
            refresh(&pb, &term);
            delay = now - prev;

            /* take screenshot */
            apply_colormap(&pb, img);
            controlgif(gsdata, -1, delay / 10000, 0, 0);
            prev = now;
            putgif(gsdata, img);
        }
        dirty = 0;
    }
    if (settings.last_frame_delay > 0) {
        controlgif(gsdata, -1, settings.last_frame_delay / 10, 0, 0);
        putgif(gsdata, img);
    }

    /* output gif */
    gifsize = endgif(gsdata);
    if (gifsize > 0) {
        write_gif(gifimage, gifsize, out_file);
        free(gifimage);
    }

    /* normal exit */
    free(img);
    if (settings.input) {
        if (fileno(in_file) != STDIN_FILENO) {
            fclose(in_file);
        }
        free(settings.input);
    }
    if (settings.output) {
        if (fileno(out_file) != STDOUT_FILENO) {
            fclose(out_file);
        }
        free(settings.output);
    }
    term_die(&term);
    pb_die(&pb);
    free(obuf);

    return nret;
}

/* emacs, -*- Mode: C; tab-width: 4; indent-tabs-mode: nil -*- */
/* vim: set expandtab ts=4 : */
/* EOF */
